rust/arrow/src/array/array.rs

```diff
@@ -1994,6 +2220,66 @@ mod tests {
         assert_eq!(2, list_array.value_length(1));
     }
 
+    #[test]
+    fn test_dictionary_array() {
+        // Construct a value array
+        let value_data = ArrayData::builder(DataType::Int8)
```

Comment:
I might need to try creating a few dictionaries so I can understand this better. I can't see where we use this array in the assertions of the test. Is the `keys` array with 3 values indexing into this one with 8 values?

Reply:
I had a look at this. I created a simple `collect_array` function, but seems like I can't access the keys and values of the dictionary as arrays directly.

```rust
impl DictionaryArray {
    /// Construct an ArrayRef from the values of the dictionary
    pub fn collect_array(&self) -> Result<ArrayRef> {
        // here I'm trying to make an array from the keys
        let keys = make_array(self.data());
        // if I have keys, cast them to u32 as they could be any other type
        // u32 -> u32 is a no-op
        let keys = crate::compute::cast(&keys, &DataType::UInt32)?;
        // get the keys as a PrimitiveArray<UIn32Type>
        let keys = keys.as_any().downcast_ref::<UInt32Array>().unwrap();
        // create an array from the values and keys
        crate::compute::take(&self.values(), &keys, None)
    }
} 
```

The above doesn't work because I couldn't figure out how to get the key and value `ArrayRef`s